### ISSUE TITLE
ttnn.hardswish_bw takes the wrong branch at exactly x = ±3 and uses 0.3333 for 1/3

### ISSUE THREAD

**Issue by ping-dev-ui** [tenstorrent/tt-metal#55291]
**Component / Area:** ttnn eltwise unary backward (composite), `hardswish_bw`

**Issue Type:** Bad Outputs

## Summary

`hardswish_bw` is composed as
([unary_backward.cpp#L923-L945](https://github.com/tenstorrent/tt-metal/blob/9fa00fce13cda1538f0130695576408ff5b05665/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L923-L945)):

```cpp
where(lt(input, -3.0f), 0.f,
      where(le(input, 3.0f), grad * (input * 0.3333f + 0.5f), grad))
```

PyTorch's reference kernel
([aten/src/ATen/native/cpu/Activation.cpp, hardswish_backward_kernel](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cpu/Activation.cpp))
is

```cpp
if (self <= -3) return 0;
else if (self < 3) return grad * (self / 3 + 0.5);
else return grad;
```

Both boundary comparisons are inverted relative to the reference: ttnn is `<` at -3 where torch is `<=`, and `<=` at +3
where torch is `<`. The registered golden for this op is torch autograd, so the op and its golden disagree at exactly
`x = ±3`. Those values are bf16-exact and ordinary in activation tensors, so exact hits are not rare.

A secondary defect on the same line: the slope constant is `0.3333f` while the forward `hardswish` kernel uses the
correctly rounded `1/6` (`hardsigmoid_init` loads `0.1666666716337204f`). In fp32 this costs 336 to 2097 ULP everywhere in
`(-3, 3)`; at `x = 2.5` the op returns `1.33325` for `1.3333333`.

This is a source-level analysis with a host replay; I do not have a card. The comparisons and constants are exact, so the
device result at `±3` is the same as the replay.

## Observed vs expected (identical in fp32 and bf16, since ±3 is exact in both)

| x | grad | `hardswish_bw` now | torch autograd | abs error |
|---|---|---|---|---|
| -3 | 1 | -0.4999 | 0 | 0.50 |
| +3 | 1 | 1.4999 | 1 | 0.50 |
| -3 | 2 | -0.9998 | 0 | 1.00 |
| +3 | -2 | -2.9998 | -2 | 1.00 |
| 2.5 | 1 | 1.33325 | 1.3333333 | 8.3e-5 |

At `x = -3` the returned gradient even has the wrong sign (negative where the reference is exactly 0).

## Why the tests do not catch it

The nightly test draws inputs from `uniform(-100, 100)` in bf16 and compares by PCC; two wrong lanes out of a
320×384 tensor do not move the PCC, and the sampled inputs rarely land on exactly ±3.

## Reproducer (host only, `pip install torch`)

```python
import torch

def ttnn_hardswish_bw(grad, x):                                   # unary_backward.cpp L923-945
    inner = grad * (x * torch.tensor(0.3333) + 0.5)
    return torch.where(x < -3, torch.tensor(0.0), torch.where(x <= 3, inner, grad))

x = torch.tensor([-3., 3., -3., 3., 2.5], requires_grad=True)
g = torch.tensor([1., 1., 2., -2., 1.])
torch.nn.functional.hardswish(x).backward(g)
print(ttnn_hardswish_bw(g, x.detach()).tolist())   # [-0.4999, 1.4999, -0.9998, -2.9998, 1.33325]
print(x.grad.tolist())                             # [0.0, 1.0, 0.0, -2.0, 1.3333333]
```

## Proposed fix

```cpp
Tensor grad_result = where(
    ttnn::le(input, -3.0f, std::nullopt, output_mem_config),
    0.f,
    where(
        ttnn::lt(input, 3.0f, std::nullopt, output_mem_config),
        ttnn::multiply(grad, ttnn::add(ttnn::multiply(input, 1.0f / 3.0f), 0.5f)),   // 0x3eaaaaab
        grad));
```

Also fix the stale comment above the function, and add `x = ±3` rows (fp32 and bf16) plus an fp32 accuracy check inside
`(-3, 3)` to `test_backward_hardswish.py`.

## Acceptance criteria (proposed)

- `ttnn.hardswish_bw` returns 0 at `x = -3` and `grad` at `x = 3` for fp32 and bf16, matching torch autograd.
- Inside `(-3, 3)` the fp32 result matches `grad * (x / 3 + 0.5)` to a few ULP (no `0.3333f` truncation).
- Nightly test extended with the boundary rows.

**Comment by a40014859-gif** [comment]
I'd like to take this bounty. I can fix the ±3 branch boundaries and replace the truncated 0.3333f slope with an exact 1/3 expression, then add focused fp32/bf16 regression coverage for the boundary values and interior accuracy. Could you assign #55291 to me?

**Comment by harmonygemini8-collab** [comment]
Requesting assignment.

🔧 Plan: Two boundary inversions vs the torch reference: L927 `lt(input,-3)` → `le` (torch: `self <= -3 → 0`), L930 `le(input,3)` → `lt` (torch: `self < 3 → linear`). And replace the `input * 0.3333f` with true division by 3 (torch uses `self/3`; 0.3333f injects up to ~1e-4 scale error). Tests at exactly x = ±3 (bf16-exact, ordinary in activation tensors) plus randomized golden.

📂 Files: `unary_backward.cpp` hardswish_bw (~L923-945) + boundary unit tests.

⏱️ ETA: 1 day. 💳 Payment: USDC (Base)

I have read and agree to the [Bounty Program Terms](https://docs.tenstorrent.com/bounty_terms.html).

**Comment by ZacLou** [comment]
I have completed root-cause analysis for this issue. The fix: correct the boundary comparisons at x = ±3 (lt → le at -3, le → lt at +3) to match PyTorch hardswish_backward, replace the truncated 0.3333f with input / 3.0f for exact 1/3. Will add fp32/bf16 regression coverage at boundary values and interior accuracy checkpoints. Could you assign #55291 to me?

**Comment by ZacLou** [comment]
Opened a PR that implements the fix described above: #55299.

**Comment by ZacLou** [comment]
Hi maintainers, the fix is ready in PR #55299. Could you please:`n1. Trigger the CI / approve the workflows for this first-time contribution.`n2. Confirm the bounty payout method and currency for #55291.`n`nHappy to address any review feedback promptly.

**Comment by ping-dev-ui** [comment]
Reporter here. Fix with tests is ready on my fork (https://github.com/ping-dev-ui/tt-metal/tree/fix/hardswish-bw-boundaries): boundary comparisons flipped to match PyTorch's hardswish_backward_kernel, exact 1.0f / 3.0f, stale comment fixed, and a nightly boundary test at x = ±3 in fp32 and bf16. Holding the PR until assignment per the bounty terms. @jberkowitzTT could you assign this to me when it is reviewed?

**Comment by ZacLou** [comment]
I have read and agree to the [Tenstorrent Bounty Program Terms & Conditions](https://docs.tenstorrent.com/bounty_terms.html). PR #55299 is already open with the fix and regression tests. Could you please assign this issue to me so the bounty is tracked correctly? Thanks.

**Comment by ping-dev-ui** [comment]
I have read and agree to the Tenstorrent Bounty Program Terms & Conditions (https://docs.tenstorrent.com/bounty_terms.html).

**Comment by jberkowitzTT** [comment]
Hi @ping-dev-ui - please submit all bounty proposals to [bounties@tenstorrent.com](mailto:bounties@tenstorrent.com). If approved, we'll create the issue and assign it to you as a bounty

Thank you for your enthusiasm & understanding
